src/Microsoft.VisualStudio.Services.Agent/JobServerQueue.cs

```diff
@@ -602,10 +609,32 @@ private async Task UploadFile(UploadFileInfo file)
                     // Create the log
                     var taskLog = await _jobServer.CreateLogAsync(_scopeIdentifier, _hubName, _planId, new TaskLog(String.Format(@"logs\{0:D}", file.TimelineRecordId)), default(CancellationToken));
 
-                    // Upload the contents
-                    using (FileStream fs = File.Open(file.Path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
+                    if (_writeToBlobstorageService)
                     {
-                        var logUploaded = await _jobServer.AppendLogContentAsync(_scopeIdentifier, _hubName, _planId, taskLog.Id, fs, default(CancellationToken));
+                        BlobIdentifierWithBlocks blobBlockId = null;
+                        using (FileStream fs = File.Open(file.Path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
+                        {
+                            try {
+                                blobBlockId = await _jobServer.UploadLogToBlobstorageService(fs, _hubName, _planId, taskLog.Id);
+
+                                int lineCount = File.ReadLines(file.Path).Count();
+
+                                // Notify TFS
+                                await _jobServer.AssociateLogAsync(_scopeIdentifier, _hubName, _planId, taskLog.Id, blobBlockId, lineCount, default(CancellationToken));
+                            }
+                            catch {
+                                // Fall back to FCS
+                                await _jobServer.AppendLogContentAsync(_scopeIdentifier, _hubName, _planId, taskLog.Id, fs, default(CancellationToken));
```

Comment:
Do we need to reset the stream back to the beginning here? I'm not 100% on how streams work but if `UploadLogToBlobstoreService` reads some data and then fails we want to include all of it in this call.

Reply:
I think you're right